crates/ty_python_semantic/src/types/tuple.rs

```diff
@@ -0,0 +1,819 @@
+//! Defines types for describing fixed- and variable-length tuples.
+//!
+//! At runtime, a Python tuple is a fixed-length immutable list of values. There is no restriction
+//! on the types of the elements of a tuple value. In the type system, we want to model both
+//! "heterogeneous" tuples that have elements of a fixed sequence of specific types, and
+//! "homogenous" tuples that have an unknown number of elements of the same single type. And in
+//! fact, we want to model tuples that are a combination of the two, with a heterogeneous prefix
+//! and/or suffix, and a homogeneous portion of unknown length in between those.
+
+use itertools::Either;
+
+use crate::types::class::{ClassType, KnownClass};
+use crate::types::{Type, TypeMapping, TypeRelation, TypeVarInstance, TypeVarVariance, UnionType};
+use crate::util::subscript::{Nth, OutOfBoundsError, PyIndex, PySlice, StepSizeZeroError};
+use crate::{Db, FxOrderSet};
+
+/// # Ordering
+/// Ordering is based on the tuple's salsa-assigned id and not on its elements.
+/// The id may change between runs, or when the tuple was garbage collected and recreated.
+#[salsa::interned(debug)]
+#[derive(PartialOrd, Ord)]
+pub struct TupleType<'db> {
+    #[returns(ref)]
+    pub(crate) tuple: Tuple<'db>,
+}
+
+impl<'db> Type<'db> {
+    pub(crate) fn tuple(db: &'db dyn Db, tuple: TupleType<'db>) -> Self {
+        // If a fixed-length (i.e., mandatory) element of the tuple is `Never`, then it's not
+        // possible to instantiate the tuple as a whole. (This is not true of the variable-length
+        // portion of the tuple, since it can contain no elements.)
+        if tuple.tuple(db).fixed_elements().any(|ty| ty.is_never()) {
+            return Type::Never;
+        }
```

Comment:
This means that we iterate over a tuple's fixed elements twice when constructing it: once in the `TupleType` construction, and once here. We could probably avoid iterating over the tuple twice by doing this check in `VariableLengthTuple::mixed()` and `FixedLengthTuple::from_elements()`, then change this check here into a `debug_assert!()` that ensures that the inner constructors upheld the invariant.

Reply:
[Here](https://github.com/astral-sh/ruff/blob/45cd117ea4a361db9041304462bca7b8be7a8a80/crates/ty_python_semantic/src/types/unpacker.rs#L281-L310) is where we're building up a new tuple spec (but not a tuple type) during unpacking. Right now that only works on fixed-length tuples, but in a later PR I want to extend that to support mixed tuples as well, and use that whole machinery for splatting. We'd use tuple specs to describe the positional parameters and arguments, and I wanted to be able to retain the information about which particular params/args are `Never` without having to collapse the whole thing down to a single `Never`.

Reply:
Or put another way, `Tuple` (the Rust type) will not only be used to represent instances of `tuple` (the Python type). That might have been too implicit in this PR? I can add some clarifying documentation about that point if so.

(I've been using "tuple spec" in my comments to mean "a sequence of types, which might have a homogeneous variable-length bit in the middle". A `tuple` instance is _one_ of the things that is described by a tuple spec, but not the only one. We do want the "collapse the `Never`s" behavior for `tuple` instances, but (I contend) that's not universally true of tuple specs.)

Reply:
oops, I posted my response before seeing your second comment. Yes, some clarifying docs would be great! And maybe it would even be good to rename your `Tuple` struct to `TupleSpec`?